README.md
# go-piping-sshd
SSH server from anywhere with [Piping Server](https://github.com/nwtgck/piping-server)

## Thanks

SSH server implementation for shell is highly based on <https://github.com/jpillora/sshd-lite/blob/master/server/server.go>. Thanks!

## Install for macOS
```bash
brew install nwtgck/piping-sshd/piping-sshd
```

## Install for Ubuntu
```bash
wget https://github.com/nwtgck/go-piping-sshd/releases/download/v0.8.0/piping-sshd-0.8.0-linux-amd64.deb
dpkg -i piping-sshd-0.8.0-linux-amd64.deb
```

Get more executables in the [releases](https://github.com/nwtgck/go-piping-sshd/releases).

## Usage

Run the command below in machine A which will be controlled.

```bash
piping-sshd mypath --password=changeme
```

Run the command below in another machine B which will control the machine A.

```bash
curl -NsS https://ppng.io/mypath/sc | socat TCP-LISTEN:2022,reuseaddr - | curl -NsST - https://ppng.io/mypath/cs
```

Run the command below in the machine B to ssh. The password is `changeme`.

```bash
ssh-keygen -R [localhost]:2022; ssh -p 2022 dummy@localhost
```

## Help

```txt
SSH server from anywhere with Piping Server

Usage:
  piping-sshd [flags]

Flags:
      --allow-empty-password      Allows to run SSH server with empty password
      --dns-server string         DNS server (e.g. 1.1.1.1:53)
  -H, --header stringArray        HTTP header
  -h, --help                      help for piping-sshd
      --http-read-buf-size int    HTTP read-buffer size in bytes (default 4096)
      --http-write-buf-size int   HTTP write-buffer size in bytes (default 4096)
  -k, --insecure                  Allow insecure server connections when using SSL
  -p, --password string           SSH user password
  -s, --server string             Piping Server URL (default "https://ppng.io")
      --shell string              Shell
  -u, --user string               SSH user name
  -v, --version                   show version
      --yamux                     Multiplex connection by yamux
```

## CAUTION

Running this command exposes SSH server to public Piping Server. Someone who knows paths on the Piping Server can run any command on the machine. 

Here are ways for using securely.

- Specify `--password` or `-p` in long or complex password.
- The paths should be long or complex.
- Use self-host Piping Server in private network
- Run this command on a machine which everyone may access to.
- Use Piping Server with some authentication methods such as [basic auth](https://github.com/nwtgck/piping-server-basic-auth-docker-compose) or [JWT](https://github.com/nwtgck/jwt-piping-server).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)

## [Unreleased]

## [0.8.0] - 2022-06-27
### Added
* Remote port forwarding

### Changed
* Update dependencies

## [0.7.0] - 2022-01-28
### Added
* Support Windows partially (Port-forwarding and SFTP supported, pty not supported)

### Changed
* (breaking) Rename --allows-empty-password to --allow-empty-password because `--allow-` is more commonly used

## [0.6.0] - 2022-01-27
### Changed
* Allow no Content-Type
* Use `--allows-empty-password` to allow the empty password

## [0.5.0] - 2022-01-25
### Added
* Support "exec" request type
* Support SFTP

### Changed
* Use an embed private key, not random one
* Use 4096 as default buffer sizes in HTTP for better speed
* Update dependencies
* Add `,reuseaddr` to the socat command	hint

## [0.4.0] - 2021-03-27
### Changed
* Make the hint command shorter
* No auth when password is empty

## [0.3.1] - 2021-03-25
### Changed
* (internal) Use github.com/creack/pty instead of github.com/kr/pty because moved

## [0.3.0] - 2021-03-25
### Added
* Support "direct-tcpip", which is for port forwarding with -L or tunneling with -D

## [0.2.0] - 2021-03-25
### Added
* Add --shell option to specify shell

## 0.1.0 - 2021-03-25
### Added
* Initial release

[Unreleased]: https://github.com/nwtgck/go-piping-sshd/compare/v0.8.0...HEAD
[0.8.0]: https://github.com/nwtgck/go-piping-sshd/compare/v0.7.0...v0.8.0
[0.7.0]: https://github.com/nwtgck/go-piping-sshd/compare/v0.6.0...v0.7.0
[0.6.0]: https://github.com/nwtgck/go-piping-sshd/compare/v0.5.0...v0.6.0
[0.5.0]: https://github.com/nwtgck/go-piping-sshd/compare/v0.4.0...v0.5.0
[0.4.0]: https://github.com/nwtgck/go-piping-sshd/compare/v0.3.1...v0.4.0
[0.3.1]: https://github.com/nwtgck/go-piping-sshd/compare/v0.3.0...v0.3.1
[0.3.0]: https://github.com/nwtgck/go-piping-sshd/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/nwtgck/go-piping-sshd/compare/v0.1.0...v0.2.0


cmd/root.go
package cmd

import (
	"fmt"
	"github.com/hashicorp/yamux"
	"github.com/nwtgck/go-piping-sshd/piping_util"
	"github.com/nwtgck/go-piping-sshd/priv_key"
	"github.com/nwtgck/go-piping-sshd/ssh_server"
	"github.com/nwtgck/go-piping-sshd/util"
	"github.com/nwtgck/go-piping-sshd/version"
	"github.com/pkg/errors"
	"github.com/spf13/cobra"
	"golang.org/x/crypto/ssh"
	"io"
	"log"
	"net/http"
	"os"
)

const (
	ServerUrlEnvName = "PIPING_SERVER"
)

var serverUrl string
var insecure bool
var dnsServer string
var showsVersion bool
var headerKeyValueStrs []string
var httpWriteBufSize int
var httpReadBufSize int
var sshYamux bool
var sshUser string
var allowsEmptyPassword bool
var sshPassword string
var sshShell string

func init() {
	cobra.OnInitialize()
	defaultServer, ok := os.LookupEnv(ServerUrlEnvName)
	if !ok {
		defaultServer = "https://ppng.io"
	}
	RootCmd.PersistentFlags().StringVarP(&serverUrl, "server", "s", defaultServer, "Piping Server URL")
	RootCmd.PersistentFlags().StringVar(&dnsServer, "dns-server", "", "DNS server (e.g. 1.1.1.1:53)")
	// NOTE: --insecure, -k is inspired by curl
	RootCmd.PersistentFlags().BoolVarP(&insecure, "insecure", "k", false, "Allow insecure server connections when using SSL")
	RootCmd.PersistentFlags().StringArrayVarP(&headerKeyValueStrs, "header", "H", []string{}, "HTTP header")
	RootCmd.PersistentFlags().IntVarP(&httpWriteBufSize, "http-write-buf-size", "", 4096, "HTTP write-buffer size in bytes")
	RootCmd.PersistentFlags().IntVarP(&httpReadBufSize, "http-read-buf-size", "", 4096, "HTTP read-buffer size in bytes")
	RootCmd.PersistentFlags().BoolVarP(&showsVersion, "version", "v", false, "show version")
	RootCmd.PersistentFlags().StringVarP(&sshUser, "user", "u", "", "SSH user name")
	RootCmd.PersistentFlags().StringVarP(&sshPassword, "password", "p", "", "SSH user password")
	RootCmd.PersistentFlags().BoolVarP(&allowsEmptyPassword, "allow-empty-password", "", false, "Allows to run SSH server with empty password")
	RootCmd.PersistentFlags().StringVarP(&sshShell, "shell", "", "", "Shell")
	RootCmd.PersistentFlags().BoolVarP(&sshYamux, "yamux", "", false, "Multiplex connection by yamux")
}

var RootCmd = &cobra.Command{
	Use:          os.Args[0],
	Short:        "piping-sshd",
	Long:         "SSH server from anywhere with Piping Server",
	SilenceUsage: true,
	RunE: func(cmd *cobra.Command, args []string) error {
		if showsVersion {
			fmt.Println(version.Version)
			return nil
		}
		if sshPassword == "" {
			if !allowsEmptyPassword {
				return fmt.Errorf("specify non-empty --password or --allow-empty-password")
			}
		}

		clientToServerPath, serverToClientPath, err := generatePaths(args)
		if err != nil {
			return err
		}
		headers, err := piping_util.ParseKeyValueStrings(headerKeyValueStrs)
		if err != nil {
			return err
		}
		httpClient := util.CreateHttpClient(insecure, httpWriteBufSize, httpReadBufSize)
		if dnsServer != "" {
			// Set DNS resolver
			httpClient.Transport.(*http.Transport).DialContext = util.CreateDialContext(dnsServer)
		}
		serverToClientUrl, err := util.UrlJoin(serverUrl, serverToClientPath)
		if err != nil {
			return err
		}
		clientToServerUrl, err := util.UrlJoin(serverUrl, clientToServerPath)
		if err != nil {
			return err
		}
		// Print hint
		sshPrintHintForClientHost(clientToServerUrl, serverToClientUrl, clientToServerPath, serverToClientPath)

		// (base: https://gist.github.com/jpillora/b480fde82bff51a06238)
		sshConfig := &ssh.ServerConfig{
			//Define a function to run when a client attempts a password login
			PasswordCallback: func(c ssh.ConnMetadata, pass []byte) (*ssh.Permissions, error) {
				// Should use constant-time compare (or better, salt+hash) in a production setting.
				if (sshUser == "" || c.User() == sshUser) && string(pass) == sshPassword {
					return nil, nil
				}
				return nil, fmt.Errorf("password rejected for %q", c.User())
			},
			// No auth when password is empty
			NoClientAuth: sshPassword == "",
		}
		// TODO: specify priv_key by flags
		pri, err := ssh.ParsePrivateKey([]byte(priv_key.PrivateKeyPem))
		if err != nil {
			return err
		}
		sshConfig.AddHostKey(pri)

		// If not using multiplexer
		if !sshYamux {
			duplex, err := piping_util.DuplexConnect(httpClient, headers, serverToClientUrl, clientToServerUrl)
			if err != nil {
				return err
			}
			// Before use, a handshake must be performed on the incoming net.Conn.
			sshConn, chans, reqs, err := ssh.NewServerConn(util.NewDuplexConn(duplex), sshConfig)
			if err != nil {
				log.Printf("Failed to handshake (%s)", err)
				return nil
			}

			log.Printf("New SSH connection from (%s)", sshConn.ClientVersion())
			go ssh_server.HandleGlobalRequests(sshConn, reqs)
			// Accept all channels
			ssh_server.HandleChannels(sshShell, chans)
		}

		// If yamux is enabled
		if sshYamux {
			log.Printf("Multiplexing with yamux")
			return sshHandleWithYamux(sshConfig, httpClient, headers, clientToServerUrl, serverToClientUrl)
		}
		return nil
	},
}

func sshPrintHintForClientHost(clientToServerUrl string, serverToClientUrl string, clientToServerPath string, serverToClientPath string) {
	clientHostPort := 2022
	if !sshYamux {
		fmt.Println("=== Client host (socat + curl) ===")
		fmt.Printf(
			"  curl -NsS %s | socat TCP-LISTEN:%d,reuseaddr - | curl -NsST - %s\n",
			serverToClientUrl,
			clientHostPort,
			clientToServerUrl,
		)
	}
	flags := ""
	if sshYamux {
		flags += fmt.Sprintf("--%s ", yamuxFlagLongName)
	}
	fmt.Println("=== Client host (piping-tunnel) ===")
	fmt.Printf(
		"  piping-tunnel -s %s client -p %d %s%s %s\n",
		serverUrl,
		clientHostPort,
		flags,
		clientToServerPath,
		serverToClientPath,
	)
	fmt.Println("=== SSH client ===")
	userAndHost := "localhost"
	if sshUser != "" {
		userAndHost = sshUser + "@localhost"
	}
	fmt.Printf("  ssh-keygen -R [localhost]:%d; ssh -p %d %s\n", clientHostPort, clientHostPort, userAndHost)
}

func sshHandleWithYamux(sshConfig *ssh.ServerConfig, httpClient *http.Client, headers []piping_util.KeyValue, clientToServerUrl string, serverToClientUrl string) error {
	var duplex io.ReadWriteCloser
	duplex, err := piping_util.DuplexConnectWithHandlers(
		func(body io.Reader) (*http.Response, error) {
			return piping_util.PipingSend(httpClient, headersWithYamux(headers), serverToClientUrl, body)
		},
		func() (*http.Response, error) {
			res, err := piping_util.PipingGet(httpClient, headers, clientToServerUrl)
			if err != nil {
				return nil, err
			}
			contentTypes := res.Header.Values("Content-Type")
			// NOTE: No Content-Type is for curl user
			// NOTE: application/octet-stream is for compatibility
			if !(len(contentTypes) == 0 || contentTypes[0] == yamuxMimeType || contentTypes[0] == "application/octet-stream") {
				return nil, errors.Errorf("invalid content-type: %s", contentTypes)
			}
			return res, nil
		},
	)
	yamuxSession, err := yamux.Server(duplex, nil)
	if err != nil {
		return err
	}
	for {
		yamuxStream, err := yamuxSession.Accept()
		if err != nil {
			return err
		}
		go func() {
			// (base: https://gist.github.com/jpillora/b480fde82bff51a06238)
			// Before use, a handshake must be performed on the incoming net.Conn.
			sshConn, chans, reqs, err := ssh.NewServerConn(yamuxStream, sshConfig)
			if err != nil {
				log.Printf("Failed to handshake (%s)", err)
				return
			}
			log.Printf("New SSH connection from %s (%s)", sshConn.RemoteAddr(), sshConn.ClientVersion())
			go ssh_server.HandleGlobalRequests(sshConn, reqs)
			// Accept all channels
			ssh_server.HandleChannels(sshShell, chans)
		}()
	}
}
